geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/rebalance/model/PartitionedRegionLoadModel.java

```diff
@@ -435,10 +435,24 @@ private void initLowRedundancyBuckets() {
 
   private void initOverRedundancyBuckets() {
     this.overRedundancyBuckets = new TreeSet<>(REDUNDANCY_COMPARATOR);
+    Set<String> redundancyZones = new HashSet<>();
+
     for (BucketRollup b : this.buckets) {
-      if (b != null && b.getOnlineRedundancy() > this.requiredRedundancy) {
-        this.overRedundancyBuckets.add(b);
+      if (b != null) {
+        if (b.getOnlineRedundancy() > this.requiredRedundancy) {
+          this.overRedundancyBuckets.add(b);
+        } else {
+          for (Member member : b.getMembersHosting()) {
+            String redundancyZone = this.getRedundancyZone(member.getDistributedMember());
+            if (redundancyZones.contains(redundancyZone)) {
```

Comment:
This check is always false, because `redundancyZones` starts off empty and is only added to if this check returns true, so there's no way for anything to ever be added to it. Negating this check results in the tests failing, so it seems that this isn't just a typo, but with this check always being false, the logic in this method becomes identical to that in the unmodified method.

In addition to this, reverting the changes in this file results in the added tests still passing, which would imply that the tests aren't testing the intended behaviour.

Reply:
I will look it this. I specifically tested the fixes and tests so, I believe your test results, but I am not sure how I ended up at this point.  Thanks for you always detailed feedback!

Reply:
So part of this change set is to roll back some changes that will perform poorly in cases not currently tested. So, it is not surprising that reverting all the changes would yield passing tests. With regards to the chunk of code you were referencing, you are correct there needed to be an else with the zone being added to redundancyZones there.

Lastly, there is also a challenge of the tests in RebalanceOperationComplexPart2DistributedTests.java not really behaving properly. The failure relies the partitioned region load model doing the wrong thing which it only does periodically depending on how it allocates buckets. I am reviewing the fact that I cannot get them to fail in the way I expect at this point. 

This also leads me to think that a unit test might yield a more consistent result than a dunit. I will look at it.

Reply:
I have rewritten most of the tests as JUnit tests, as well as fixing the a couple of issues I found in the code. Please review when you get the chance.